### DISCUSSION TITLE
"Toggle" global api authentification via env var without creating migrations

### DISCUSSION THREAD

**Discussion by seriousm4x** [pocketbase/pocketbase#1967] [Q&A]
I'm planning to add an environment variable like `AUTH_ENABLED=true/false` to protect all api routes or none at all.

The only way I've found is to change the access permissions on collection level for every rule. Meaning, this is how I've implemented it:

```go
func StartPocketBase() {
	App = pocketbase.New()

	App.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		var err error
		authEnabled := false
		authEnabledEnv := os.Getenv("AUTH_ENABLED")
		if authEnabledEnv != "" {
			authEnabled, err = strconv.ParseBool(authEnabledEnv)
			if err != nil {
				logger.Error.Println(err)
			}
		}

		if err := setApiRules(authEnabled); err != nil {
			return err
		}

		return nil
	})

	if err := App.Start(); err != nil {
		logger.Error.Fatalln(err)
	}
}

func setApiRules(authEnabled bool) error {
	var apiRule *string
	unauthorizedRule := ""
	apiRule = &unauthorizedRule
	if authEnabled {
		apiRule = nil
	}

	// set devices api rule
	devicesCollection, err := App.Dao().FindCollectionByNameOrId("devices")
	if err != nil {
		return err
	}
	devicesCollection.CreateRule = apiRule
	devicesCollection.DeleteRule = apiRule
	devicesCollection.ListRule = apiRule
	devicesCollection.UpdateRule = apiRule
	devicesCollection.ViewRule = apiRule
	if err := App.Dao().SaveCollection(devicesCollection); err != nil {
		return err
	}

	// set ports api rule
	portsCollection, err := App.Dao().FindCollectionByNameOrId("ports")
	if err != nil {
		return err
	}
	portsCollection.CreateRule = apiRule
	portsCollection.DeleteRule = apiRule
	portsCollection.ListRule = apiRule
	portsCollection.UpdateRule = apiRule
	portsCollection.ViewRule = apiRule
	if err := App.Dao().SaveCollection(portsCollection); err != nil {
		return err
	}

	// set settings api rule
	settingsCollection, err := App.Dao().FindCollectionByNameOrId("settings")
	if err != nil {
		return err
	}
	settingsCollection.CreateRule = apiRule
	settingsCollection.DeleteRule = apiRule
	settingsCollection.ListRule = apiRule
	settingsCollection.UpdateRule = apiRule
	settingsCollection.ViewRule = apiRule
	if err := App.Dao().SaveCollection(settingsCollection); err != nil {
		return err
	}

	return nil
}
```

The problem with this is, that it will create new migrations every time the env var changes.
When I push database changes to my git repo, users migration might fail because of this.

Is there a simpler way to "toggle" global api authentification?

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand your use case and how the env toggle is supposed to work.

What will be the expected result if the user has `AUTH_ENABLED=false` and then decide to change the API rule from the UI or the API? Should the collection rule be resetted on the next server start? Eventually, it is possible to prevent collection API rules change if you attach to the `OnModelBefore*` or `OnCollectionBefore*Request` hooks, but still the behavior is ambiguous.

In any case, if you want to avoid triggering `Dao` events you can create a new `Dao` instance as shown in https://pocketbase.io/docs/database/#dao-without-event-hooks:

```go
dao := daos.New(app.Dao().DB())

// use dao instead of app.Dao()...
```

---

> Is there a simpler way to "toggle" global api authentification?

I'm not sure that I understand what do you mean. You can register a [global middleware](https://pocketbase.io/docs/router/#middlewares) via the `e.Router.Use(...)` and check the env variable from there and enforce any extra constraints, but there is no global mechanism to ignore the collection API rule checks within their related action.

One hacky workaround could be to temporary "reset" or "nullify" (depending on your flag) the collection from the request context inside the middleware, but I don't understand your use case and I can't recommend it:

```go
collection, _ := c.Get(apis.ContextCollectionKey).(*models.Collection)

// nullify for admin only access
collection.ListRule = nil

// OR reset to empty string to allow everyone access
collection.ListRule = types.Pointer("")
```

**Reply by ganigeorgiev** [reply]
Also keep in mind that the collection API rules act as filters for the client-side apis (eg. when calling `getList()` as a regular user) and globally resetting the rules may have some unintended side-effects.

**Reply by seriousm4x** [reply]
> I'm not sure that I understand your use case and how the env toggle is supposed to work.

Sorry if that was unclear. I want to give users the option to choose between authorized use only or unauthorized use only.
Basically if `AUTH_ENABLED` is `true`, all `/api/*` routes should only be accessible by authorized admin users and unauthorized users would be redirect to a login form. If `false`, all routes should be accessible by anyone unauthorized.

Using `e.Router.Use()` looks like the best thing to do. Although I serve a website with static files at `/` like this:

```go
e.Router.GET("/*", apis.StaticDirectoryHandler(distDirFS, true))
```
which should always be reachable, no matter what `AUTH ENABLED` is set to. So I thought about something like this:
```go
e.Router.GET("/api/*", ???, apis.RequireAdminAuth())
```
Would that work? What would be the default HandlerFunc that pocketbase uses?

---

> One hacky workaround could be to "reset" or "nullify" (depending on your flag) the collection from the request context inside the middleware, but I don't understand your use case and I can't recommend it:

Thats what I thought as well.

**Reply by ganigeorgiev** [reply]
> e.Router.GET("/api/*", ???, apis.RequireAdminAuth())
> Would that work? What would be the default HandlerFunc that pocketbase uses?

No, this will register the middleware only for the specified handler.

You can create a routing group with `e.Router.Group()` and attach the handler to all routes from that group, but for the already registered routes this will not work.

Instead, you can register the global middleware and perform a path check and ignore the special handling.

```go
package main

import (
	"log"
	"strings"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func myCustomMiddleware(app core.App) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			if !strings.HasPrefix(c.Path(), "/api/") {
				return next(c)
			}

			// do something...

			return next(c)
		}
	}
}

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.Use(myCustomMiddleware(app))

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by seriousm4x** [reply]
Will try that. Thanks for your awesome support and fast replys
